### PR TITLE
Package MT migration skill as a portable Copilot plugin + add MT migration reviewer agent

### PR DESCRIPTION
**Pull request by jankratochvilcz**
## Why

We've migrated ~50 MSBuild tasks to the MT execution model across dotnet/msbuild and dotnet/sdk. The same defect classes keep coming up in review, and most hide behind helper calls — not in the task body. This PR captures those patterns as a portable Copilot CLI plugin reusable on any repo authoring tasks against the `IMultiThreadableTask` / `TaskEnvironment` API.

## What

New plugin at `plugins/mt-migration/`:

- **`skills/multithreaded-task-migration/SKILL.md`** — the migration playbook. Adds Sins 7 & 8 to the existing skill (exception path leakage, `Path.IsPathRooted` as absolutize gate), a Call-Chain Hazards section (helpers that capture process state, nested `new Task()` instantiation bypassing TaskFactory injection, ToolTask override hot spots, engine-owned immutable env vars), Test Patterns A/B with explicit guidance on when *not* to write a migration test, helper-plumbing API discipline, and an expanded sign-off checklist. **The previous in-repo copy at `.github/skills/multithreaded-task-migration/` is removed** — the plugin is now the single canonical home.
- **`agents/mt-migration-reviewer.agent.md`** — a focused PR reviewer that delegates the generic 24-dimension review to the host repo's expert reviewer, mandatorily traces every call chain from `Execute()` to leaves, and leaves **inline, line-anchored** review comments (one per finding, full chain named in each, GitHub `suggestion` blocks for fixes).
- **`plugin.json`** + **`README.md`** with cross-repo install instructions.

The generic `expert-reviewer` agent is intentionally untouched — the MT reviewer composes with whichever generic reviewer the host repo has.

No code changes — pure docs/tooling.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

**Comment by ViktorHofer** [comment]
Do you want to add a marketplace.json file as well to expose the plugin?

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+|--------------------------------------------------|----------------------------------------------------|
+| `Environment.GetEnvironmentVariable("VAR");`       | `TaskEnvironment.GetEnvironmentVariable("VAR");`     |
+| `Environment.SetEnvironmentVariable("VAR", "v");`  | `TaskEnvironment.SetEnvironmentVariable("VAR", "v");` |
+
+### Step 4: Replace Process Start APIs
+
+| BEFORE (UNSAFE - inherits process state)           | AFTER (SAFE - uses task's isolated environment)     |
+|----------------------------------------------------|-----------------------------------------------------|
+| `var psi = new ProcessStartInfo("tool.exe");`      | `var psi = TaskEnvironment.GetProcessStartInfo();`  |
+|                                                    | `psi.FileName = "tool.exe";`                        |
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md]: Fixed in 858afd4b0f.

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+
+## Updating Unit Tests
+
+Built-in MSBuild tasks now initialize `TaskEnvironment` with a `MultiProcessTaskEnvironmentDriver`-backed default. Tests creating instances of built-in tasks no longer need manual `TaskEnvironment` setup. For custom or third-party tasks that implement `IMultiThreadableTask` without a default initializer, set `TaskEnvironment = TaskEnvironmentHelper.CreateForTest()`.
+
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md:65]: Fixed in 858afd4b0f.

File: plugins/mt-migration/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+/plugin install file://$(pwd)/plugins/mt-migration
+```
+
+### Across multiple repos (recommended)
+
+Publish this plugin folder to a small standalone repo (e.g.
+`your-org/mt-migration-plugin`) and install from there. The plugin contains no
+hard-coded references to `dotnet/msbuild` paths — only to the public
+`Microsoft.Build.Framework` API surface — so it works on any task-authoring
+repo unchanged.
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply plugins/mt-migration/README.md]: Fixed in 858afd4b0f.

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 4]
```diff
@@ -222,6 +232,158 @@ Old code threw `FileNotFoundException` for missing files; new code throws `Argum
...
+| `ValidateParameters()` | Often calls `File.Exists` on input parameters | Absolutize before probing |
+| `GenerateResponseFileCommands()` / `GenerateCommandLineCommands()` | Response-file/command-line args are passed to the spawned process — but the process's `WorkingDirectory` is the project directory (set by `TaskEnvironment.GetProcessStartInfo()`), so relative args resolve correctly in the child. **Do not absolutize tool arguments** — it inflates user-visible output and can leak absolute paths into tool diagnostics or generated artifacts. |
+| `GetWorkingDirectory()` | Default returns `null` → child inherits host CWD | Leave alone if you use `TaskEnvironment.GetProcessStartInfo()`; that already sets `WorkingDirectory` to project dir |
+
+**Key contract:** `ProcessStartInfo.FileName` is resolved by the OS **before** `WorkingDirectory` takes effect, so the executable path must be absolute. Tool arguments are interpreted by the *child process* in its working directory, so they should remain relative.
+
+### Environment Variables: Immutable Set in MT Mode
+
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply .github/skills/multithreaded-task-migration/SKILL.md]: Fixed in 858afd4b0f.

File: plugins/mt-migration/agents/mt-migration-reviewer.agent.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+You review pull requests that migrate (or partially migrate) MSBuild tasks to the multithreaded execution model. Your job is to find MT-specific defects that a general reviewer will miss — particularly **defects hiding behind helper calls and library boundaries** that the task code itself doesn't make visible.
+
+You **do not** re-explain the migration playbook. The playbook lives in the `multithreaded-task-migration` skill — load it, apply it, but never quote it back at the author. They have read it. Your value is in finding what they missed.
+
+---
+
+## Operating Rules
+
+1. **Delegate general review.** If the host repo has an `expert-reviewer` agent (or equivalent — e.g., `dotnet-msbuild:msbuild-code-review`), invoke it for the 24-dimension pass. Do not redo style, perf, naming, or generic concurrency. Read its output, then layer your MT-specific findings on top. Do not repeat its findings.
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply plugins/mt-migration/agents/mt-migration-reviewer.agent.md]: Fixed in 858afd4b0f.

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 6]
```diff
@@ -222,6 +232,158 @@ Old code threw `FileNotFoundException` for missing files; new code throws `Argum
...
+catch (FileNotFoundException ex) { Log.LogErrorWithCodeFromResources("MSB3484", absPath.OriginalValue); }
+
+// For ex.Message that may embed the path: sanitize by string-replacing absolute with original
+catch (Exception ex) { Log.LogError(ex.Message.Replace(absPath.Value, absPath.OriginalValue)); }
+```
+
+**Detect**: For every `Log.LogError(ex.Message …)` / `Log.LogError(…, ex.FileName)` downstream of a `GetAbsolutePath`, ask: does the exception originate from a helper that received the absolutized path? If yes, sanitize.
+
+### Sin 8: `Path.IsPathRooted` as an Absolutize Gate
+
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply .github/skills/multithreaded-task-migration/SKILL.md]: Fixed in 858afd4b0f.

**Comment by jankratochvilcz** [comment]
Added `.github/plugin/marketplace.json` to expose the plugin via the standard Copilot CLI marketplace mechanism.

**Review by ViktorHofer — APPROVED** [review]

File: .github/plugin/marketplace.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "msbuild-plugins",
+  "owner": {
+    "name": "dotnet/msbuild",
+    "email": "msbuild@microsoft.com"
```
Comment (JanProvaznik) [inline .github/plugin/marketplace.json]: 1. that's not our email
2. do we want to expose email here?

File: .github/plugin/marketplace.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  },
+  "metadata": {
+    "description": "Copilot CLI plugins for the MSBuild ecosystem",
```
Comment (JanProvaznik) [inline .github/plugin/marketplace.json]: I think copilot cli is not the only surface for this, so it's misleading

File: plugins/mt-migration/agents/mt-migration-reviewer.agent.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+You review pull requests that migrate (or partially migrate) MSBuild tasks to the multithreaded execution model. Your job is to find MT-specific defects that a general reviewer will miss — particularly **defects hiding behind helper calls and library boundaries** that the task code itself doesn't make visible.
+
+You **do not** re-explain the migration playbook. The playbook lives in the `multithreaded-task-migration` skill — load it, apply it, but never quote it back at the author. They have read it. Your value is in finding what they missed.
+
+---
+
+## Operating Rules
+
+1. **Delegate general review.** If the host repo has an `expert-reviewer` agent (or equivalent), invoke it for the 24-dimension pass. Do not redo style, perf, naming, or generic concurrency. Read its output, then layer your MT-specific findings on top. Do not repeat its findings.
```
Comment (JanProvaznik) [inline plugins/mt-migration/agents/mt-migration-reviewer.agent.md:18]: weird to say this and then publish it as marketplace

File: .github/plugin/marketplace.json (hunk 8 above)
Comment (ViktorHofer) [reply .github/plugin/marketplace.json]: This marketplace only works with Copilot so that's fine. The claude code, codex, etc. must be placed in a different file location. But the CLI term is in-accurate, that's right.

File: plugins/mt-migration/agents/mt-migration-reviewer.agent.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+2. If yes → the test is theater. MAJOR finding.
+3. The test must use either **Pattern A (decoy-CWD)** or **Pattern B (cross-instance ProjectDirectory divergence)** from the skill — anything else is unlikely to fail when the migration regresses.
+4. CWD-mutating tests must be pinned to a non-parallel xUnit collection. If not, flag as MAJOR (intermittent flake).
+5. Tests that use `Path.GetTempPath()` + `Guid.NewGuid()` + manual cleanup instead of `TestEnvironment`/`TransientTestFolder` — MINOR (leaks on failure).
+
+### Step 5 — Verify scope of the migration
+
+If the task is normally invoked via the TaskFactory system (declared in a `.tasks` file or used from targets as `<MyTask … />`), the attribute alone is sufficient *provided* the call chain is clean. Verify in the host repo's `.targets` / `.tasks` files. If the task is **only** instantiated by other tasks via `new MyTask()` (e.g., `TlbImp` inside `ResolveComReference`), the migration is mostly meaningless until the parent is migrated and propagates its `TaskEnvironment` — flag as MAJOR with a recommendation to either bundle the parent migration or close the PR.
```
Comment (AlesProkop) [inline plugins/mt-migration/agents/mt-migration-reviewer.agent.md]: I think it makes sense to migrate a task that is only invoked by other tasks. If we bundle them, the PR might get big and the task is useless but not harmful and has to be done eventually. It is definitely worth a comment, maybe even todo in the code but I would not call that a MAJOR issue.

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md [hunk 11]
```diff
@@ -222,6 +231,122 @@ Old code threw `FileNotFoundException` for missing files; new code throws `Argum
 
 **Detect**: For every `GetAbsolutePath`, check what the old code threw for null/empty and whether the calling code has type-specific catch blocks.
 
+### Sin 7: Exception Message Path Leakage
```
Comment (OvesN) [inline plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md]: This is a copy of sin 2

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md [hunk 12]
```diff
@@ -128,7 +137,7 @@ After migration, review for behavioral compatibility. **Every observable differe
 
 Observable behavior = `Execute()` return value, `[Output]` property values, error/warning message content, exception types, files written, and which code path runs.
 
-## The 6 Deadly Compatibility Sins
+## The 8 Deadly Compatibility Sins
 
 Real bugs found during MSBuild task migrations. Every one shipped in initial "passing" code with green tests.
```
Comment (OvesN) [inline plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md:142]: Based on my observations, I would also add that the agent should pay particular attention to edge cases: what happens if the input is null, an empty string, or contains only whitespace? Will the application’s behavior change after the migration?

I also like to add the following to the prompt:

“Investigate all possible scenarios, even if they appear extremely unlikely. If a given scenario is at least theoretically possible, treat it as a relevant finding worth mentioning.”

File: plugins/mt-migration/agents/mt-migration-reviewer.agent.md (hunk 9 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/agents/mt-migration-reviewer.agent.md:18]: I think we need to figure out how this performs, then we can polish it for external consumption. My immediate goal is to have this for us for cross-repo tho'

File: .github/plugin/marketplace.json (hunk 7 above)
Comment (ViktorHofer) [reply .github/plugin/marketplace.json]: Yes, we don't need an email in this marketplace.

Comment (jankratochvilcz) [reply .github/plugin/marketplace.json]: Fixed in 465f79dace — removed the email field entirely.

File: .github/plugin/marketplace.json (hunk 8 above)
Comment (jankratochvilcz) [reply .github/plugin/marketplace.json]: Fixed in 465f79dace — renamed to 'GitHub Copilot' (not CLI-specific).

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md (hunk 11 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md]: They're related but distinct: Sin 2 is about directly logging an `AbsolutePath` in error messages (e.g. `Log.LogError(..., abs)`). Sin 7 is about *exception messages* that indirectly embed the absolute path (e.g. `Log.LogError(ex.Message)` where the exception was thrown by a helper that received the absolutized path). The fix is different too — Sin 2 uses `OriginalValue`, Sin 7 requires sanitizing the exception string. I'll add a cross-reference to make the distinction clearer.

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md (hunk 12 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md:142]: Good point — null/empty/whitespace edge cases are exactly the kind of thing that changes behavior post-migration (Sin 6 touches on this with `GetAbsolutePath` throwing on empty). I'll fold explicit null/empty/whitespace checking into the agent's call-chain audit protocol in a follow-up.

File: plugins/mt-migration/agents/mt-migration-reviewer.agent.md (hunk 10 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/agents/mt-migration-reviewer.agent.md]: Fair — downgrading from MAJOR to MINOR with a recommendation to add a TODO comment makes sense. The migration isn't harmful, just incomplete without the parent. Will adjust the severity guidance.

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md (hunk 11 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md]: Added a cross-reference block in b92c95c9bd clarifying the distinction: Sin 2 = directly logging an `AbsolutePath`; Sin 7 = exception messages that *indirectly* embed the absolute path from a helper call.

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md (hunk 12 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md:142]: Addressed in b92c95c9bd — added an edge-case discipline paragraph to the deadly sins intro requiring null/empty/whitespace verification for every migrated code path.

File: plugins/mt-migration/agents/mt-migration-reviewer.agent.md (hunk 10 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/agents/mt-migration-reviewer.agent.md]: Addressed in b92c95c9bd — downgraded from MAJOR to MINOR, changed recommendation from 'bundle or close' to 'add a TODO comment + file follow-up issue'.

File: plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md (hunk 11 above)
Comment (jankratochvilcz) [reply plugins/mt-migration/skills/multithreaded-task-migration/SKILL.md]: Consolidated in 3a3800df8e — merged old Sin 7 (exception message path leakage) into Sin 2 (now 'Error/Log Message Path Inflation') covering both direct and indirect leakage. Renumbered old Sin 8 → Sin 7. Updated all cross-references.

**Review by AlesProkop — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]
